### PR TITLE
[iOS] Improve background layer frame mapping performance

### LINKED ISSUE #20218: Can't use LinearGradientBrush as Button Background on iOS
### Description

Works on Android (tested on Pixel 7 API 34 emulator)
android
but on iOS (tested on iPhone 15 Pro Max iOS 17.2 emulator)
ios
there is something wrong...

### Steps to Reproduce

1. Create a Button.
2. Set Background property to LinearGradientBrush.
3. Run on iOS.

### Link to public reproduction project repository

https://github.com/alececihs/MauiButtonsTest/

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

_No response_

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #24847: [iOS/MacCatalyst] Background layer frame mapping has poor performance
### Description

Using https://github.com/davidortinau/AllTheLists app, and scrolling through collection view with `Border` inside we can clearly see that a good amount of time is spent in `UpdateBackgroundLayerFrame` which is needed to sync the sublayer with the `UIView.Layer`.

image

This has an impact on both `MappingFrame` and `ContentView.LayoutSubviews`:
image
image
image

Setting `Frame` on a background layer shouldn't take so much time.

### Steps to Reproduce

_No response_

### Link to public reproduction project repository

_No response_

### Version with bug

8.0.90 SR9

### Is this a regression from previous behavior?

No, this is something new

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, macOS

### Affected platform versions

_No response_

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #26057: [iOS & Mac] Gradient background size is incorrect when invalidating parent
### Description

When invalidating a parent, gradient applied to the child fills the entire screen.

In my example, I have a Grid with 2 Rows(40,*) and loaded a button as first row with LinearGradientBrush.
When calling Grid.InvalidateMeasure(), the gradient size fills the entire screen.

### Steps to Reproduce

1. Run the attached sample.
2. Click anywhere outside the Button.
[Gradient.zip](https://github.com/user-attachments/files/17870319/Gradient.zip)

### Link to public reproduction project repository

_No response_

### Version with bug

9.0.10 SR1

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, macOS

### Affected platform versions

_No response_

### Did you find any workaround?

Yes
Rename the gradient relevant layer name to skip size updating for this particaluar layer alone.
    private void ButtonLoaded(object? sender, EventArgs e)
    {
        #if IOS
            if(sender is Button button && button.Handler != null && button.Handler.PlatformView != null && button.Handler.PlatformView is UIKit.UIView platformView)
            {
                if(platformView.Layer != null && platformView.Layer.Sublayers != null)
                {
                    if(platformView.Layer != null && platformView.Layer.Sublayers != null)
                {
                    var layer = platformView.Layer.Sublayers.FirstOrDefault(x => x.Name == "MauiBackgroundLayer");
                    if( layer != null && layer is CALayer gradientLayer)
                    {
                        gradientLayer.Name = "GradientLayer";
                    }
                }
                }
            }
#endif

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by albyrock87**
### Important

This PR `Obsolete`s a couple of methods.
```
static Microsoft.Maui.Controls.Platform.BrushExtensions.UpdateBackgroundLayer(this UIKit.UIView view) -> void
static Microsoft.Maui.Platform.ViewExtensions.UpdateBackgroundLayerFrame(this UIKit.UIView! view) -> void
```

### Description of Change

Using https://github.com/davidortinau/AllTheLists app, and scrolling through collection view with `Border` inside we can clearly see that a good amount of time is spent in `UpdateBackgroundLayerFrame` which is needed to sync the sublayer with the `UIView.Layer`.

image

This has an impact on both `MappingFrame` and `ContentView.LayoutSubviews`:
image

Which translated to these timings
image
image

This PR gets rid of all mapping frame code and simply adds an observer on the sublayer.

As a result we get a lot of improvement (80%):
image
image

While the new observer is super fast
image

### Issues Fixed

Fixes #24847
Fixes #26057
Fixes #20218
Closes #26222

### CODE REVIEW

**Comment by albyrock87** [comment]
Failing tests definitely unrelated :)

File: src/Controls/src/Core/Platform/iOS/Extensions/BrushExtensions.cs [hunk 1]
```diff
@@ -176,9 +178,6 @@ public static void RemoveBackgroundLayer(this CALayer layer)
 			}
 		}
 
-		public static void UpdateBackgroundLayer(this UIView view) =>
```
Comment (rmarinho) [inline src/Controls/src/Core/Platform/iOS/Extensions/BrushExtensions.cs:179]: We can. t remove this on a servicing , this would need to target net9.0

**Review by rmarinho — DISMISSED** [review]
I think this will need to retarget net9.0 , right @PureWeen  ?

**Comment by albyrock87** [comment]
I'll take the time to install .NET9 preview and retarget this one on top of that branch.

**Comment by albyrock87** [comment]
@rmarinho @PureWeen I've rebased this onto `.net9.0` and retargeted the PR.
May you run /AZP again? thanks

**Comment by albyrock87** [comment]
Hold on on this one, I have to verify something.

**Comment by albyrock87** [comment]
So I've come across random crashes in DeviceTests in `base.RemoveFromSuperview()` invocation and I found out that the problem was strangely caused by the observer.

After an in depth investigation I think I've found the root cause of the problem which I strongly believe lies in [`Action<NSObservedChange> cback;`](https://github.com/xamarin/xamarin-macios/blob/main/src/Foundation/NSObject2.cs#L1059) not being a weak reference too.

This brought me to implement a low-level solution which is now safe (I can't reproduce the crash anymore).
https://github.com/dotnet/maui/pull/24848/commits/d8763bfb933a10ea8ba40cb3f872d8068e7b5e62

File: src/Compatibility/Core/src/iOS/Renderers/FrameRenderer.cs [hunk 2]
```diff
@@ -136,7 +136,6 @@ public override void LayoutSubviews()
 			if (_previousSize != Bounds.Size)
 			{
 				SetNeedsDisplay();
-				this.UpdateBackgroundLayer();
```
Comment (jsuarezruiz) [inline src/Compatibility/Core/src/iOS/Renderers/FrameRenderer.cs:139]: Could include an UITest? Should be a VerticalStackLayout including a Frame and a Border with gradients background and changing the View size at runtime. If the gradients renders correctly after change the size, the test pass. To change the size, can use something like a Button. And to verify the rendering, could use the `VerifyScreenshot` method.


Let me know if needs help or want myself to include it.

**Review by jsuarezruiz — DISMISSED** [review]

File: src/Compatibility/Core/src/iOS/Renderers/FrameRenderer.cs (hunk 2 above)
Comment (albyrock87) [reply src/Compatibility/Core/src/iOS/Renderers/FrameRenderer.cs:139]: @jsuarezruiz I've added the UI test as requested, feel free to add any additional test if you feel it's needed.

**Comment by albyrock87** [comment]
I've added a device test to address memory leak / crash concerns.

**Review by PureWeen — CHANGES_REQUESTED** [review]

**Comment by albyrock87** [comment]
Failed test is unrelated.

**Comment by albyrock87** [comment]
I've included @Tamilarasan-Paranthaman UI test from https://github.com/dotnet/maui/pull/26222

File: src/Compatibility/Core/src/iOS/Renderers/PageRenderer.cs [hunk 3]
```diff
@@ -135,9 +135,6 @@ public override void ViewDidLayoutSubviews()
 
 			if (_safeAreasSet || !Forms.IsiOS11OrNewer)
 				UpdateUseSafeArea();
-
-			if (Element.Background != null && !Element.Background.IsEmpty)
-				NativeView?.UpdateBackgroundLayer();
```
Comment (PureWeen) [inline src/Compatibility/Core/src/iOS/Renderers/PageRenderer.cs:140]: Can you rebase and just remove the changes from the Compatibility project?

We could even move versions of UpdateBackgroundLayerFrame into compatibility
Once we delete them from controls

We try to never touch any of the code inside `Compatibility`

**Review by PureWeen — CHANGES_REQUESTED** [review]

**Review by PureWeen — APPROVED** [review]

**Comment by PureWeen** [comment]
- failing test on iOS/Catalyst is a known issue on main and not related to this PR

**Pull request merged**